tracee-rules/benchmark/signature/wasm/signatures.go

```diff
@@ -0,0 +1,202 @@
+package wasm
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+
+	"github.com/open-policy-agent/opa/rego"
+
+	"github.com/aquasecurity/tracee/tracee-rules/types"
+	"github.com/open-policy-agent/golang-opa-wasm/opa"
+)
+
+var (
+	//go:embed helpers.rego
+	helpersRego string
+
+	//go:embed anti_debugging_ptraceme.rego
+	antiDebuggingPtracemeRego string
+
+	//go:embed code_injection.rego
+	codeInjectionRego string
+)
+
+type CodeInjectionWASMSignature struct {
+	types.Signature
+	rego *opa.OPA
+	cb   types.SignatureHandler
+}
+
+func compileRegoToWasm(regoCodes []string) []byte {
+	re := regexp.MustCompile(`package\s.*`) // TODO: DRY
+
+	var pkgName string
+	regoMap := make(map[string]string)
+	var compiledRego *ast.Compiler
+
+	for _, regoCode := range regoCodes {
+		regoModuleName := strings.Split(re.FindString(regoCode), " ")[1]
+		if !strings.Contains(regoCode, "package tracee.helpers") {
+			pkgName = regoModuleName
+		}
+		regoMap[regoModuleName] = regoCode
+	}
+
+	var err error
+	compiledRego, err = ast.CompileModules(regoMap)
+	if err != nil {
+		panic(err)
+	}
+
+	//fmt.Println(pkgName)
+
+	cr, err := rego.New(
+		rego.Compiler(compiledRego),
```

Comment:
I tried it but the results are rather unexpectedly slower.

```
 go test -tags=opa_wasm -bench=. -benchmem -benchtime=100x | prettybench -no-passthrough
benchmark                                     iter           time/iter     bytes alloc             allocs
---------                                     ----           ---------     -----------             ------
BenchmarkWASMOptionTarget_SimpleRule-2         100      80832.00 ns/op      13086 B/op      372 allocs/op
BenchmarkWASMLibrary_SimpleRule-2              100      11549.00 ns/op       1678 B/op       30 allocs/op
BenchmarkRegoGoLibrary_SimpleRule-2            100       9814.00 ns/op       7964 B/op      123 allocs/op
BenchmarkWASMOptionTarget_RealRule-2           100     122719.00 ns/op      25430 B/op      664 allocs/op
BenchmarkWASMLibrary_RealRule-2                100      20513.00 ns/op       1984 B/op       30 allocs/op
BenchmarkRegoGoLibrary_RealRule-2              100      38014.00 ns/op      18290 B/op      379 allocs/op
```

Notice the build flag of `opa_wasm` must be passed in for it to use the WASM runtime. By default it is disabled. Not sure why that is. 

Regardless I will still push these new tests for completeness.

Reply:
Latest released OPA version (v0.29.4):
```
go test -tags=opa_wasm -bench=. -benchmem -benchtime=100x | prettybench -no-passthrough
benchmark                                     iter           time/iter     bytes alloc             allocs
---------                                     ----           ---------     -----------             ------
BenchmarkWASMOptionTarget_SimpleRule-2         100     316014.00 ns/op      34621 B/op     1251 allocs/op
BenchmarkWASMLibrary_SimpleRule-2              100       9334.00 ns/op       1677 B/op       30 allocs/op
BenchmarkRegoGoLibrary_SimpleRule-2            100       8391.00 ns/op       7908 B/op      122 allocs/op
BenchmarkWASMOptionTarget_RealRule-2           100     343139.00 ns/op      47059 B/op     1553 allocs/op
BenchmarkWASMLibrary_RealRule-2                100      23539.00 ns/op       1984 B/op       30 allocs/op
BenchmarkRegoGoLibrary_RealRule-2              100      30570.00 ns/op      18069 B/op      377 allocs/op
```

Reply:
I thought that we'll just add BenchmarkOnEventCodeInjectionRuleWASM and BenchmarkEngineWithCodeInjectionRuleWASM to the family:
BenchmarkOnEventCodeInjectionRuleRego
BenchmarkOnEventCodeInjectionRuleGo
BenchmarkEngineWithCodeInjectionRuleRego
BenchmarkEngineWithCodeInjectionRuleWASM

This will allow us to relate benchmarks that use the same (or comparable) rule and inputs